### ISSUE TITLE
lossy quantization

### ISSUE THREAD

**Issue by ZechenM** [ZechenM/distributed-ml-training#12]
zero out all the floating points for gradients and make them all integers

For future, do binary search to figure out the best precision (i.e. the optimal number of floating points) that balance network latency and model accuracy

**Comment by ZechenM** [comment]
lossy quantization would not work

1. zeroing out all the decimal points directly makes all gradients 0
2. cutting a floating-point number from 32-bit to 8-bit has the following issue

e.g. 0.0009765625 is much computationally easier to represent than 0.3 in IEEE 754 (if both Python and TCP are adopting IEEE 754 to represent floating-point numbers)

**Comment by ZixiQu** [comment]
"zeroing out all the decimal points directly makes all gradients 0; cutting a floating-point number from 32-bit to 8-bit has the following issue"

Excellent finding. I guess we can close this issue?

**Comment by ZechenM** [comment]
for 1, still yes
for 2, I found out that what Pytorch claims that cutting 64 bits to 32 bits/16bits is just using different IEEE 754 precisions. When they say they are quantizing 64/32 bits to 32/16 bits floating point precisions, they are using IEEE 754 half/single precision to represent a decimal number that was originally in single/double precision 

will close this issue with the above findings.
